src/nifs/protogalaxy/poly.rs

```diff
@@ -0,0 +1,179 @@
+use std::num::NonZeroU32;
+
+use ff::Field;
+use halo2curves::CurveAffine;
+use itertools::Itertools;
+use rayon::prelude::*;
+use tracing::*;
+
+use crate::{
+    fft,
+    plonk::{self, eval, GetChallenges, GetWitness, PlonkStructure},
+    polynomial::{lagrange, univariate::UnivariatePoly},
+};
+
+use super::pow_i;
+
+#[derive(Debug, thiserror::Error, PartialEq, Eq, Clone)]
+pub enum Error {
+    #[error(transparent)]
+    Eval(#[from] eval::Error),
+    #[error("pow_i iterator ended before witnesses")]
+    PowiEndedBeforeWitnesses,
+}
+
+/// Analog of [`itertools::structs::MultiProduct`], but:
+/// - without [`Clone`] require
+/// - without logic for diff len of iterators
+struct MultiProduct<Item, Iter>
+where
+    Iter: Iterator<Item = Item>,
+{
+    iterators: Box<[Iter]>,
+}
+
+impl<Item, Iter> Iterator for MultiProduct<Item, Iter>
+where
+    Iter: Iterator<Item = Item>,
+{
+    type Item = Box<[Item]>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iterators.iter_mut().map(|p| p.next()).collect()
+    }
+}
+
+#[instrument(skip_all)]
+pub(crate) fn compute_F<C: CurveAffine>(
+    beta: C::ScalarExt,
+    delta: C::ScalarExt,
+    S: &PlonkStructure<C::ScalarExt>,
+    trace: &(impl Sync + GetChallenges<C::ScalarExt> + GetWitness<C::ScalarExt>),
+) -> Result<UnivariatePoly<C::ScalarExt>, Error> {
+    let count_of_rows = 2usize.pow(S.k as u32);
+    let count_of_gates = S.gates.len();
+
+    let count_of_evaluation = count_of_rows * count_of_gates;
+
+    let points_count = count_of_evaluation
+        .next_power_of_two()
+        .ilog2()
+        .next_power_of_two() as usize;
+
+    let log_n = NonZeroU32::new(points_count.ilog2()).unwrap();
+
+    debug!(
+        "
+        count_of_rows: {count_of_rows};
+        count_of_gates: {count_of_gates};
+        count_of_evaluation: {count_of_evaluation};
+        log_n: {log_n};
+        points_count: {points_count}"
+    );
+
+    let mut evaluated_points = MultiProduct {
+        iterators: lagrange::iter_cyclic_subgroup::<C::ScalarExt>(log_n.get())
+            .map(|X| pow_i::iter_eval_of_pow_i(log_n, beta + X * delta))
+            .take(points_count)
+            .collect::<Box<[_]>>(),
+    }
+    .zip(plonk::iter_evaluate_witness::<C>(S, trace))
+    .map(
+        |(mut evaluated_pow_i_per_challenge, result_with_evaluated_w_i)| {
+            let evaluated_w_i = result_with_evaluated_w_i?;
+
+            evaluated_pow_i_per_challenge
+                .iter_mut()
+                .for_each(|evaluated_pow_i| {
+                    *evaluated_pow_i *= evaluated_w_i;
```

Comment:
`What TODO? We calculate one point by O(n), but we have log(n) points`

IIUC, you calculate:  `sum_i pow_i(...)*f_i(w)` it involves the evaluation of pow_i for each i. There is a better algorithm described in the paper that don't have to evaluate pow_i for each i. I also described it in the diagram in the complexity analysis.

Reply:
Of course there is no description of the algorithm, it's just one picture without annotations, but I guessed about your idea, I'll do it today

Reply:
That picture is to illustrate protogalaxy paper Claim 4.4. In that claim, there is detailed description to compare. However, the graph is more clear than the description.

Reply:
My bad, I forgot about this part, I implemented the algorithm in a single-threaded environment. 

For multithreaded environment I couldn't find tree_reduce, and also couldn't find try_tree_reduce (so that after the first error everything stops executing).

I propose to stop the single-threaded version now, for multithreading is a separate volume task

Reply:
"stop the single-threaded version now", do you mean we stop at the single-threaded version and pause for the multi-thread version? Yeah, we can make a TODO for multi-thread version. I don't have a clear picture to do the multi-threading as well.

Reply:
Yes, I have already added TODO about `try_*` (more optimal error handling, no overhead) and paralleling. About the latter - I have a clear plan for implementation, however, the task is large and I need to read more about [rayon's parallel iterators](https://github.com/rayon-rs/rayon/blob/main/src/iter/plumbing/README.md) to estimate the task